pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -51,107 +52,112 @@ type kubeController struct {
 	}
 	mc.initSecretController(kc)
 
 	return mc, nil
 }
 
+func (m *Multicluster) EnableCertPatch(caBundlePath string, fetchCaRoot func() map[string]string) {
+	m.caBundlePath = caBundlePath
+	m.fetchCaRoot = fetchCaRoot
+}
+
 // AddMemberCluster is passed to the secret controller as a callback to be called
 // when a remote cluster is added.  This function needs to set up all the handlers
 // to watch for resources being added, deleted or changed on remote clusters.
-func (m *Multicluster) AddMemberCluster(clients kubelib.Client, clusterID string) error {
+func (m *Multicluster) AddMemberCluster(client kubelib.Client, clusterID string) error {
 	// stopCh to stop controller created here when cluster removed.
 	stopCh := make(chan struct{})
 	var remoteKubeController kubeController
 	remoteKubeController.stopCh = stopCh
 	m.m.Lock()
-	options := Options{
-		SystemNamespace:   m.systemNamespace,
-		WatchedNamespaces: m.WatchedNamespaces,
-		ResyncPeriod:      m.ResyncPeriod,
-		DomainSuffix:      m.DomainSuffix,
-		XDSUpdater:        m.XDSUpdater,
-		ClusterID:         clusterID,
-		NetworksWatcher:   m.networksWatcher,
-		Metrics:           m.metrics,
-		EndpointMode:      m.endpointMode,
-		SyncInterval:      m.syncInterval,
-	}
+	options := m.opts
+	options.ClusterID = clusterID
 	log.Infof("Initializing Kubernetes service registry %q", options.ClusterID)
-	kubectl := NewController(clients, options)
+	kubeRegistry := NewController(client, options)
 
-	remoteKubeController.Controller = kubectl
-	m.serviceController.AddRegistry(kubectl)
+	remoteKubeController.Controller = kubeRegistry
+	m.serviceController.AddRegistry(kubeRegistry)
 
 	m.remoteKubeControllers[clusterID] = &remoteKubeController
 	m.m.Unlock()
 
 	// Only need to add service handler for kubernetes registry as `initRegistryEventHandlers`,
 	// because when endpoints update `XDSUpdater.EDSUpdate` has already been called.
-	_ = kubectl.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+	_ = kubeRegistry.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+
+	// TODO move instance cache out of registries
+	if m.serviceEntryStore != nil && features.EnableServiceEntrySelectPods {
+		// Add an instance handler in the kubernetes registry to notify service entry store about pod events
+		_ = kubeRegistry.AppendWorkloadHandler(m.serviceEntryStore.WorkloadInstanceHandler)
+	}
+
+	if m.serviceEntryStore != nil && features.EnableK8SServiceSelectWorkloadEntries {
+		// Add an instance handler in the service entry store to notify kubernetes about workload entry events
+		_ = m.serviceEntryStore.AppendWorkloadHandler(kubeRegistry.WorkloadInstanceHandler)
+	}
 
-	go kubectl.Run(stopCh)
+	go kubeRegistry.Run(stopCh)
 	webhookConfigName := strings.ReplaceAll(validationWebhookConfigNameTemplate, validationWebhookConfigNameTemplateVar, m.secretNamespace)
 	if m.fetchCaRoot != nil {
```

Comment:
This logic to only create the namespace controller for remote clusters shouldn't be necessary. We can consolidate `initNamespaceController` + `initSidecarInjector` as well. Might be different for the cert patch though.. not sure why these things are coupled together.. seems like`caBundlePath` corresponds to patch cert, and `fetchCaRoot` is for namespace controller. 

Also:

https://github.com/istio/istio/blob/d49931e2e89093cefcfe2ab58d928ed190539bd5/pilot/pkg/bootstrap/server.go#L1037

do we need to do leader election for the multicluster namespace controllers as well? In a multi-primary setup we're currently having all primaries perform this for all other clusters (regardless if the other clusters are primary or remote).  

cc @irisdingbj can you provide some clarity?

Reply:
We do need leader protect for this as well. Since it is there, i do not point it here. Just filed a bug yesterday. https://github.com/istio/istio/issues/28779